accepted/00065-salt-minions-update-cycle.md

```diff
@@ -0,0 +1,127 @@
+
+Rationale: Track all the patches in only one place. As the code is isolated (see above), avoid unnecessary extra-packaging for isolated version of the Salt Minion.
+
+## Principle
+
+The principle is to run own minimal [PyPi](https://pypi.org/project/pypiserver/) local server on Uyuni cluster. In order to install a Salt Minion, a managed system should use PIP againstlocal PyPi server, as shown on Figure 1 below:
+
+![Layout](images/00065-layout.png)
+
+_Figure 1_
+
+Process update or installation flow as follows:
+1. One RPM general package of Salt Minion **(1)** is provided for all possible platforms and architectures and is installed on PyPi server. However, Salt Minion is not _installed_ per se in the sense as for end-user, but as a PyPi package into the PyPi repository location. This way patches are still tracked at RPM `.spec` level.
+2. General statically compiled Python vanilla environment is provided as an RPM package from Uyuni repository and is allowed to be installed **(2)** on a Managed System, e.g. in `/opt/python3.8` or whatever version is needed. That way it is also allowed to have many Python versions as well as its modifications, if ever needed.
+3. Once virtual environment (not shown on the _Figure 1_) is created for the the separate Python interpreter, PIP is bootstrapping the entire Salt Minion against Uyuni-served PyPi server with all the required Salt Minion dependencies.
+
+This should result into running fully featured Salt Minion of latest version in an isolated virtual environment on an OS-isolated Python interpreter, which is aligned with the current Uyuni Server. Once Uyuni Server is updated to another version, so is the PyPi container image is also upgraded. Thus, Uyuni Server should issue mass-upgrade to all Salt Minions that should PIP-upgrade themselves, bumping up to the latest available version.
+
+Consequently, this mechanism will allow to keep always aligned Salt Minions across the entire infrastructure consistent and identical.
+
+## Fail-proof Updates
+
+Having running Salt Minion inside a virtual environment, allows to deliver fail-proof updates. The basic mechanism is to clone Python Virtual Environment from the current active Salt Minion and update the Salt Minion there.
+
+If update is finished successfully, Salt Minion is restarted into a new environment, which is set as default. Previos environment is kept, and is purged over one update cycle (one after current update). This way administrator always has an ability to roll-back to the previous known working version.
+
+If update failed, Salt Minion keeps running without changes, report is created and new Virtual Environment is purged immeditately.
+
+As shown on the _Figure 2_, the difference between RPM-based updates is that the software components are overwritten, instead of installed nearby. Therefore once Salt Minion is broken on upgrade for whatever reasons, it might be no longer possible to revert this with the faulty Salt Minion itself.
```

Comment:
Using a venv as a poor-man container is dangerous, and you need to implement of the logic behind the orchestration, upgrade, replacement and rollback.

Reply:
No, venv is not any dangerous and is not a "poor-man container" either. It is just different python path to load modules, and nothing else. Set shell's variables different and you have different modules to load for the Python interpreter. The same identical way works PyInstaller, which you are advocating for.

Reply:
venv is for development, no for production. Is brittle, full of escapes that can break the system, the scripts and services needs to be patched in place, etc. Is a really nightmare when you use them outside development.

> The same identical way works PyInstaller, which you are advocating for.

Did I?